config/load.go

```diff
@@ -0,0 +1,298 @@
+		}
+	}
+
+	// load environment variables
+	v.SetEnvPrefix(cfg.AppName)
+	v.AllowEmptyEnv(true)
+	v.AutomaticEnv()
+
+	appPrefix := cfg.AppName
+	if appPrefix != "" {
+		appPrefix += "."
+	}
+
+	flags := getFlagRefs(cmd)
+
+	for _, configuration := range configurations {
+		configureViper(cfg, v, reflect.ValueOf(configuration), flags, appPrefix, "")
+
+		// unmarshal fully populated viper object onto config
+		err := v.Unmarshal(configuration)
+		if err != nil {
+			return err
+		}
+
+		// Convert all populated config options to their internal application values ex: scope string => scopeOpt source.Scope
+		err = processPostConfig(configuration)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// configureViper loads the default configuration values into the viper instance, before the config values are read and parsed
+func configureViper(cfg Config, v *viper.Viper, value reflect.Value, flags flagRefs, appPrefix string, path string) {
+	if value.Type().Kind() == reflect.Ptr && value.Type().Elem().Kind() != reflect.Struct {
+		if flag, ok := flags[value.Pointer()]; ok {
+			cfg.Log.Trace(fmt.Sprintf("binding: %s = %v (flag)\n", strings.ToUpper(regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(appPrefix+path, "_")), value.Elem().Interface()))
+			_ = v.BindPFlag(path, flag)
+			return
+		}
+	}
+
+	if value.Type().Kind() == reflect.Ptr {
+		value = value.Elem()
+	}
+
+	if value.Type().Kind() != reflect.Struct {
+		cfg.Log.Trace(fmt.Sprintf("binding: %s = %v\n", strings.ToUpper(regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(appPrefix+path, "_")), value.Interface()))
+		v.SetDefault(path, value.Interface())
+		return
+	}
+
+	if path != "" {
+		path += "."
+	}
+
+	// for each field in the configuration struct, see if the field implements the defaultValueLoader interface and invoke it if it does
+	for i := 0; i < value.NumField(); i++ {
+		fieldValue := value.Field(i)
+		field := value.Type().Field(i)
+
+		mapStructTag := field.Tag.Get("mapstructure")
+
+		if mapStructTag == "-" {
+			continue
+		}
+
+		if !field.Anonymous && mapStructTag == "" {
+			cfg.Log.Trace(fmt.Sprintf("not binding field due to lacking mapstructure tag: %s.%s", value.Type().Name(), field.Name))
+			continue
+		}
+
+		if fieldValue.Type().Kind() != reflect.Ptr {
+			fieldValue = fieldValue.Addr()
+		}
+
+		configureViper(cfg, v, fieldValue, flags, appPrefix, path+mapStructTag)
+	}
+}
+
+type flagRefs map[uintptr]*pflag.Flag
+
+func getFlagRefs(cmd *cobra.Command) flagRefs {
+	refs := flagRefs{}
+	for _, flags := range []*pflag.FlagSet{cmd.PersistentFlags(), cmd.Flags()} {
+		flags.VisitAll(func(flag *pflag.Flag) {
+			v := reflect.ValueOf(flag.Value)
+			// check for struct types like stringArrayValue
+			if v.Type().Kind() == reflect.Ptr {
+				vf := v.Elem()
+				if vf.Type().Kind() == reflect.Struct {
+					if _, ok := vf.Type().FieldByName("value"); ok {
+						vf = vf.FieldByName("value")
+						if vf.IsValid() {
```

Comment:
nit / style: this is pretty nested, is there a better way to write this function to be a little flatter?

Reply:
There is more than one way to make this simpler, but decomposing this into a separate function doesn't make it more complicated, it makes it simpler to read:
```golang
func getFlagRefs(cmd *cobra.Command) flagRefs {
	refs := flagRefs{}
	for _, flags := range []*pflag.FlagSet{cmd.PersistentFlags(), cmd.Flags()} {
		flags.VisitAll(func(flag *pflag.Flag) {
			refs[normalizeFlagReference(flag)] = flag
		})
	}
	return refs
}

func normalizeFlagReference(flag *pflag.Flag) uintptr {
	v := reflect.ValueOf(flag.Value)

	// check for struct types like stringArrayValue
	if v.Type().Kind() == reflect.Ptr {
		vf := v.Elem()
		if vf.Type().Kind() == reflect.Struct {
			if _, ok := vf.Type().FieldByName("value"); ok {
				vf = vf.FieldByName("value")
				if vf.IsValid() {
					v = vf
				}
			}
		}
	}
	return v.Pointer()
}
```